src/core/registration.c

```diff
@@ -24,6 +24,89 @@
 #include "registration.c.clog.h"
 #endif
 
+void
+QuicRegistrationClose(
+    _Inout_ QUIC_REGISTRATION* Registration
+    )
+{
+    QuicTraceEvent(
+        RegistrationCleanup,
+        "[ reg][%p] Cleaning up",
+        Registration);
+
+    if (Registration->CloseCompleteHandler == NULL) {
+        CxPlatEventSet(Registration->CloseEvent);
+        CxPlatThreadWait(&Registration->CloseThread);
+        CxPlatThreadDelete(&Registration->CloseThread);
+    }
+
+    if (Registration->ExecProfile != QUIC_EXECUTION_PROFILE_TYPE_INTERNAL) {
+        CxPlatLockAcquire(&MsQuicLib.Lock);
+        CxPlatListEntryRemove(&Registration->Link);
+        CxPlatLockRelease(&MsQuicLib.Lock);
+    }
+
+    CxPlatRundownReleaseAndWait(&Registration->Rundown);
+
+    QuicWorkerPoolUninitialize(Registration->WorkerPool);
+    CxPlatRundownUninitialize(&Registration->Rundown);
+    CxPlatDispatchLockUninitialize(&Registration->ConnectionLock);
+    CxPlatLockUninitialize(&Registration->ConfigLock);
+    CxPlatEventUninitialize(Registration->CloseEvent);
+
+    if (Registration->ExecProfile != QUIC_EXECUTION_PROFILE_TYPE_INTERNAL) {
+        CxPlatLockAcquire(&MsQuicLib.Lock);
+        if (MsQuicLib.CustomExecutions && CxPlatListIsEmpty(&MsQuicLib.Registrations)) {
+            //
+            // To allow all references to the execution context to be released,
+            // clean up implicitly allocated internal resources when the final
+            // registration is closed.
+            //
+            // Alternatively, MsQuic could expose an explicit ExecutionShutdown
+            // API that completes asynchronously and requires the exeuction
+            // contexts to continue executing for the duration. This is slightly
+            // awkward to impose upon apps, though, because these resources were
+            // implicitly created.
+            //
+            MsQuicLibraryLazyUninitialize();
+
+        }
+        CxPlatLockRelease(&MsQuicLib.Lock);
+    }
+}
+
+CXPLAT_THREAD_CALLBACK(RegistrationCloseWorker, Context)
+{
+    QUIC_REGISTRATION* Registration = Context;
+
+    CxPlatEventWaitForever(Registration->CloseEvent);
```

Comment:
This can spuriously wake so you'll want this to be in a loop checking to make sure it's time to clean up and going back to waiting on the event when it's not.

Reply:
Woah, really? Does "spuriously wake" mean the wait forever returns without CloseEvent being set?

Reply:
I was thinking of the APC thing you addressed in your other comment.